--- a/python/magic.py
+++ b/python/magic.py
@@ -127,14 +127,14 @@
 class Magic(object):
     def __init__(self, ms):
         self._magic_t = ms
-	self._close = _close
+        self._close = _close
 
     def close(self):
         """
         Closes the magic database and deallocates any resources used.
         """
-        if _close:
-            _close(self._magic_t)
+        if self._close:
+            self._close(self._magic_t)
 
     @staticmethod
     def __tostr(s):
