--- nss/lib/freebl/genload.c.const-qual-fix	2025-11-25 15:52:31.669711787 +0100
+++ nss/lib/freebl/genload.c	2025-11-25 15:53:03.103908883 +0100
@@ -85,7 +85,7 @@ loader_LoadLibInReferenceDir(const char*
 {
     PRLibrary* dlh = NULL;
     char* fullName = NULL;
-    char* c;
+    const char* c;
     PRLibSpec libSpec;
 
     /* Remove the trailing filename from referencePath and add the new one */
