From 2638b86aad48a3d25704caac9d7e37368fd87815 Mon Sep 17 00:00:00 2001
From: Semyon Moroz <donbarbos@proton.me>
Date: Tue, 18 Nov 2025 23:45:35 +0400
Subject: [PATCH] Remove `no_type_check_decorator` from `__all__` for Python >=
 3.15 (#699)

---
 src/typing_extensions.py | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/typing_extensions.py b/src/typing_extensions.py
index ca1705e4..9dd8eac4 100644
--- a/src/typing_extensions.py
+++ b/src/typing_extensions.py
@@ -148,7 +148,6 @@
     'ValuesView',
     'cast',
     'no_type_check',
-    'no_type_check_decorator',
 ]
 
 # for backward compatibility
@@ -4312,3 +4311,7 @@ def type_repr(value):
 Generic = typing.Generic
 ForwardRef = typing.ForwardRef
 Annotated = typing.Annotated
+
+# Breakpoint: https://github.com/python/cpython/pull/133602
+if sys.version_info < (3, 15, 0):
+    __all__.append("no_type_check_decorator")
