From 2ad3aca11e60a9f50574a65df2232dfb58597f2b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Zbigniew=20J=C4=99drzejewski-Szmek?= <zbyszek@in.waw.pl>
Date: Mon, 12 Jan 2026 13:57:06 +0100
Subject: [PATCH 2/6] dlopen-notes: group two helper functions together

---
 dlopen-notes.py | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/dlopen-notes.py b/dlopen-notes.py
index e55800c98d..f103611c00 100755
--- a/dlopen-notes.py
+++ b/dlopen-notes.py
@@ -19,6 +19,11 @@ try:
 except ImportError:
     print_json = print
 
+def dictify(f):
+    def wrap(*args, **kwargs):
+        return dict(f(*args, **kwargs))
+    return functools.update_wrapper(wrap, f)
+
 def listify(f):
     def wrap(*args, **kwargs):
         return list(f(*args, **kwargs))
@@ -63,11 +68,6 @@ class ELFFileReader:
 
                 yield from j
 
-def dictify(f):
-    def wrap(*args, **kwargs):
-        return dict(f(*args, **kwargs))
-    return functools.update_wrapper(wrap, f)
-
 @dictify
 def group_by_soname(elffiles):
     for elffile in elffiles:
