diff -ur gettext-0.26.orig/gettext-tools/gnulib-tests/test-stdcountof-h.c gettext-0.26/gettext-tools/gnulib-tests/test-stdcountof-h.c
--- gettext-0.26.orig/gettext-tools/gnulib-tests/test-stdcountof-h.c	2025-07-08 14:24:40.000000000 +0530
+++ gettext-0.26/gettext-tools/gnulib-tests/test-stdcountof-h.c	2026-01-13 13:07:35.226196031 +0530
@@ -42,17 +42,21 @@
 
   (void) local_bounded;
 
+#ifdef _gl_verify_is_array
   (void) _gl_verify_is_array (unbounded);
   (void) _gl_verify_is_array (bounded);
   (void) _gl_verify_is_array (multidimensional);
+#endif
 
   ASSERT (countof (bounded) == 10);
   ASSERT (countof (multidimensional) == 10);
   ASSERT (countof (local_bounded) == 20);
 
 #if 0 /* These produce compilation errors.  */
+#ifdef _gl_verify_is_array
   (void) _gl_verify_is_array (integer);
   (void) _gl_verify_is_array (parameter);
+#endif
 
   ASSERT (countof (integer) >= 0);
   ASSERT (countof (unbounded) >= 0);
