From 2d1641d76c0254e3c95f461d9a0fd080e0f29c3c Mon Sep 17 00:00:00 2001
From: Jakub Martisko <jamartis@redhat.com>
Date: Fri, 23 Jan 2026 13:21:41 +0100
Subject: [PATCH] increase tr count

---
 lib/dfa.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/dfa.c b/lib/dfa.c
index 0b694e1..d76e9ec 100644
--- a/lib/dfa.c
+++ b/lib/dfa.c
@@ -354,7 +354,7 @@ typedef struct
 
 /* Maximum for any transition table count.  This should be at least 3,
    for the initial state setup.  */
-enum { MAX_TRCOUNT = 1024 };
+enum { MAX_TRCOUNT = 2048 };
 
 /* A bracket operator.
    e.g., [a-c], [[:alpha:]], etc.  */
-- 
2.52.0

