From b164c6a329464cb7f0de160f52d0468130cbd4e8 Mon Sep 17 00:00:00 2001
From: Roman Inflianskas <rominf@pm.me>
Date: Wed, 14 May 2025 12:44:46 +0300
Subject: [PATCH] test(logs): avoid failures when running with integrations
 enabled

---
 tests/test_logs.py | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/tests/test_logs.py b/tests/test_logs.py
index 1f6b07e7..cf67d933 100644
--- a/tests/test_logs.py
+++ b/tests/test_logs.py
@@ -448,6 +448,8 @@ def test_logger_with_all_attributes(sentry_init, capture_envelopes):
     assert isinstance(attributes["thread.name"], str)
     del attributes["thread.name"]
 
+    assert attributes.pop("sentry.sdk.name").startswith("sentry.python")
+
     # Assert on the remaining non-dynamic attributes.
     assert attributes == {
         "foo": "bar",
@@ -458,7 +460,6 @@ def test_logger_with_all_attributes(sentry_init, capture_envelopes):
         "sentry.message.template": "log #%d",
         "sentry.message.parameters.0": 1,
         "sentry.environment": "production",
-        "sentry.sdk.name": "sentry.python",
         "sentry.sdk.version": VERSION,
         "sentry.severity_number": 13,
         "sentry.severity_text": "warn",
-- 
2.49.0

