diff -u --recursive hugo-0.159.0-vanilla/modules/client_test.go hugo-0.159.0/modules/client_test.go
--- hugo-0.159.0-vanilla/modules/client_test.go	2026-03-23 13:16:59.000000000 -0500
+++ hugo-0.159.0/modules/client_test.go	2026-03-23 21:53:48.417518709 -0500
@@ -34,6 +34,7 @@
 )
 
 func TestClient(t *testing.T) {
+	t.Skip("skip test as network access is not allowed during package build")
 	htesting.SkipSlowTestUnlessCI(t)
 	modName := "hugo-modules-basic-test"
 	modPath := "github.com/gohugoio/tests/" + modName
