diff -up bison-2.7/src/getargs.c\~ bison-2.7/src/getargs.c
--- bison-2.7/src/getargs.c~	2012-12-05 14:51:18.000000000 +0100
+++ bison-2.7/src/getargs.c	2013-05-17 17:45:53.188616509 +0200
@@ -80,7 +80,6 @@ int skeleton_prio = default_prio;
 const char *skeleton = NULL;
 int language_prio = default_prio;
 struct bison_language const *language = &valid_languages[0];
-const char *include = NULL;
 
 
 /** Decode an option's set of keys.
@@ -500,12 +499,10 @@ static char const short_options[] =
   "b:"
   "d"
   "f::"
-  "e"
   "g::"
   "h"
   "k"
   "l"
-  "n"
   "o:"
   "p:"
   "r:"
@@ -535,7 +532,6 @@ static struct option const long_options[
 
   /* Parser. */
   { "name-prefix",   required_argument,	  0,   'p' },
-  { "include",       required_argument,   0,   'I' },
 
   /* Output. */
   { "file-prefix", required_argument,	0,   'b' },
@@ -564,7 +560,6 @@ static struct option const long_options[
   { "force-define",   required_argument,         0,   'F' },
   { "locations",      no_argument,		 0, LOCATIONS_OPTION },
   { "no-lines",       no_argument,               0,   'l' },
-  { "raw",            no_argument,               0,     0 },
   { "skeleton",       required_argument,         0,   'S' },
   { "language",       required_argument,         0,   'L' },
   { "token-table",    no_argument,               0,   'k' },
@@ -623,10 +618,6 @@ getargs (int argc, char *argv[])
         }
 	break;
 
-      case 'I':
-	include = AS_FILE_NAME (optarg);
-	break;
-
       case 'L':
 	language_argmatch (optarg, command_line_prio,
 			   command_line_location ());

Diff finished.  Fri May 17 17:45:58 2013
