diff -up lua-5.4.8/src/lparser.c.54bug3 lua-5.4.8/src/lparser.c
--- lua-5.4.8/src/lparser.c.54bug3	2026-01-14 16:43:08.796019341 -0500
+++ lua-5.4.8/src/lparser.c	2026-01-14 16:43:32.426630214 -0500
@@ -940,6 +940,8 @@ static void constructor (LexState *ls, e
     if (ls->t.token == '}') break;
     closelistfield(fs, &cc);
     field(ls, &cc);
+    checklimit(fs, cc.tostore + cc.na + cc.nh, INT_MAX/2,
+               "items in a constructor");
   } while (testnext(ls, ',') || testnext(ls, ';'));
   check_match(ls, '}', '{', line);
   lastlistfield(fs, &cc);
