diff -Naur mariadb-11.8.6/mysql-test/main/skip_grants.result mariadb-11.8.6_patched/mysql-test/main/skip_grants.result
--- mariadb-11.8.6/mysql-test/main/skip_grants.result	2026-01-31 14:27:46.000000000 +0100
+++ mariadb-11.8.6_patched/mysql-test/main/skip_grants.result	2026-02-18 13:58:02.842715556 +0100
@@ -152,4 +152,44 @@
 #
 select * from information_schema.users;
 USER	PASSWORD_ERRORS	PASSWORD_EXPIRATION_TIME
+#
+# MDEV-38811 crash in information_schema.table_constraints when --skip-grant-tables
+#
+select * from information_schema.table_constraints;
+CONSTRAINT_CATALOG	CONSTRAINT_SCHEMA	CONSTRAINT_NAME	TABLE_SCHEMA	TABLE_NAME	CONSTRAINT_TYPE
+def	mysql	Host	mysql	roles_mapping	UNIQUE
+def	mysql	Options	mysql	servers	CHECK
+def	mysql	PRIMARY	mysql	column_stats	PRIMARY KEY
+def	mysql	PRIMARY	mysql	columns_priv	PRIMARY KEY
+def	mysql	PRIMARY	mysql	db	PRIMARY KEY
+def	mysql	PRIMARY	mysql	event	PRIMARY KEY
+def	mysql	PRIMARY	mysql	func	PRIMARY KEY
+def	mysql	PRIMARY	mysql	global_priv	PRIMARY KEY
+def	mysql	PRIMARY	mysql	gtid_slave_pos	PRIMARY KEY
+def	mysql	PRIMARY	mysql	help_category	PRIMARY KEY
+def	mysql	PRIMARY	mysql	help_keyword	PRIMARY KEY
+def	mysql	PRIMARY	mysql	help_relation	PRIMARY KEY
+def	mysql	PRIMARY	mysql	help_topic	PRIMARY KEY
+def	mysql	PRIMARY	mysql	index_stats	PRIMARY KEY
+def	mysql	PRIMARY	mysql	plugin	PRIMARY KEY
+def	mysql	PRIMARY	mysql	proc	PRIMARY KEY
+def	mysql	PRIMARY	mysql	procs_priv	PRIMARY KEY
+def	mysql	PRIMARY	mysql	proxies_priv	PRIMARY KEY
+def	mysql	PRIMARY	mysql	servers	PRIMARY KEY
+def	mysql	PRIMARY	mysql	table_stats	PRIMARY KEY
+def	mysql	PRIMARY	mysql	tables_priv	PRIMARY KEY
+def	mysql	PRIMARY	mysql	time_zone	PRIMARY KEY
+def	mysql	PRIMARY	mysql	time_zone_leap_second	PRIMARY KEY
+def	mysql	PRIMARY	mysql	time_zone_name	PRIMARY KEY
+def	mysql	PRIMARY	mysql	time_zone_transition	PRIMARY KEY
+def	mysql	PRIMARY	mysql	time_zone_transition_type	PRIMARY KEY
+def	mysql	Priv	mysql	global_priv	CHECK
+def	mysql	name	mysql	help_category	UNIQUE
+def	mysql	name	mysql	help_keyword	UNIQUE
+def	mysql	name	mysql	help_topic	UNIQUE
+def	sys	PRIMARY	sys	sys_config	PRIMARY KEY
+Warnings:
+Warning	1286	Unknown storage engine 'InnoDB'
+Warning	1286	Unknown storage engine 'InnoDB'
+Warning	1286	Unknown storage engine 'InnoDB'
 # End of 11.8 tests
diff -Naur mariadb-11.8.6/mysql-test/main/skip_grants.test mariadb-11.8.6_patched/mysql-test/main/skip_grants.test
--- mariadb-11.8.6/mysql-test/main/skip_grants.test	2026-01-31 14:27:46.000000000 +0100
+++ mariadb-11.8.6_patched/mysql-test/main/skip_grants.test	2026-02-18 13:12:21.663377393 +0100
@@ -191,4 +191,10 @@
 --echo #
 select * from information_schema.users;
 
+--echo #
+--echo # MDEV-38811 crash in information_schema.table_constraints when --skip-grant-tables
+--echo #
+--sorted_result
+select * from information_schema.table_constraints;
+
 --echo # End of 11.8 tests
diff -Naur mariadb-11.8.6/sql/sql_acl.cc mariadb-11.8.6_patched/sql/sql_acl.cc
--- mariadb-11.8.6/sql/sql_acl.cc	2026-01-31 14:27:48.000000000 +0100
+++ mariadb-11.8.6_patched/sql/sql_acl.cc	2026-02-18 13:13:21.601008459 +0100
@@ -3927,6 +3927,9 @@
 privilege_t acl_get_all3(Security_context *sctx, const char *db,
                          bool db_is_patern)
 {
+  if (!initialized)
+    return DB_ACLS;
+
   privilege_t access= acl_get(sctx->host, sctx->ip,
                               sctx->priv_user, db, db_is_patern);
   if (sctx->priv_role[0])
