diff -urp libcap-ng-0.9.orig/configure.ac libcap-ng-0.9/configure.ac
--- libcap-ng-0.9.orig/configure.ac	2026-01-11 10:58:58.000000000 -0500
+++ libcap-ng-0.9/configure.ac	2026-01-26 20:28:05.975074846 -0500
@@ -123,6 +123,18 @@ if test "x$BPF_ARCH" = "x"; then
 AC_MSG_ERROR([failed to determine BPF target arch; use --with-bpf-arch])
 fi
 
+AC_MSG_CHECKING(deprecated)
+AC_ARG_ENABLE([deprecated],
+AS_HELP_STRING([--enable-deprecated],
+[enable deprecated utilities such as captest [[default=no]]]),
+[enable_deprecated=$enableval],
+[enable_deprecated=no])
+if test x$enable_deprecated = xno ; then
+	AC_MSG_RESULT(no)
+else
+	AC_MSG_RESULT(yes)
+fi
+
 AC_SUBST(BPF_ARCH)
 AC_SUBST(CLANG)
 AC_SUBST(BPFTOOL)
@@ -132,6 +144,7 @@ AC_SUBST(LIBAUDIT_CFLAGS)
 AC_SUBST(LIBAUDIT_LIBS)
 
 AM_CONDITIONAL([BUILD_CAP_AUDIT], [test "x$enable_cap_audit" = "xyes"])
+AM_CONDITIONAL([BUILD_DEPRECATED], [test "x$enable_deprecated" = "xyes"])
 
 echo .
 echo Checking for header files
diff -urp libcap-ng-0.9.orig/utils/Makefile.am libcap-ng-0.9/utils/Makefile.am
--- libcap-ng-0.9.orig/utils/Makefile.am	2026-01-11 10:58:58.000000000 -0500
+++ libcap-ng-0.9/utils/Makefile.am	2026-01-26 20:29:47.166975680 -0500
@@ -29,14 +29,19 @@ else
 SUBDIRS =
 endif
 DIST_SUBDIRS = cap-audit
-EXTRA_DIST = $(man_MANS)
+EXTRA_DIST = $(man_MANS) captest.c captest.8
 AM_CPPFLAGS = -I${top_srcdir} -I${top_srcdir}/src
 LDADD = ${top_builddir}/src/libcap-ng.la
 AM_CFLAGS = -W -Wall -Wshadow ${WFLAGS} -Wundef -D_GNU_SOURCE
-bin_PROGRAMS = pscap netcap filecap captest
-man_MANS = pscap.8 netcap.8 filecap.8 captest.8
+bin_PROGRAMS = pscap netcap filecap
+man_MANS = pscap.8 netcap.8 filecap.8
 
 pscap_SOURCES = pscap.c
 netcap_SOURCES = netcap.c proc-llist.c proc-llist.h
 filecap_SOURCES = filecap.c
+
+if BUILD_DEPRECATED
+bin_PROGRAMS += captest
+man_MANS += captest.8
 captest_SOURCES = captest.c
+endif
