diff --git a/meson.build b/meson.build
index ab28396..b5829ca 100644
--- a/meson.build
+++ b/meson.build
@@ -459,6 +459,7 @@ with_systemd = false
 systemd = dependency('systemd', required: get_option('systemd'))
 if systemd.found()
   systemduserunitdir = systemd.get_variable(pkgconfig : 'systemduserunitdir')
+  with_systemd = true
 endif
 
 configure_file(output: 'config.h', configuration: conf)
@@ -488,4 +489,4 @@ if get_option('nls')
   subdir('po')
 endif
 subdir('bash-completion')
-subdir('zsh-completion')
+# subdir('zsh-completion')
