From 880a6b48b706515442207b5409b4d2058e60146b Mon Sep 17 00:00:00 2001
From: Jakub Jelen <jjelen@redhat.com>
Date: Wed, 21 Jan 2026 17:55:25 +0100
Subject: [PATCH] tests: Workaround softhsm-2.7.0 bug in hashed ECDSA

Signed-off-by: Jakub Jelen <jjelen@redhat.com>
---
 tests/pkcs11/setup-softhsm-tokens.sh | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/tests/pkcs11/setup-softhsm-tokens.sh b/tests/pkcs11/setup-softhsm-tokens.sh
index 863c0c55e..ddc684a06 100755
--- a/tests/pkcs11/setup-softhsm-tokens.sh
+++ b/tests/pkcs11/setup-softhsm-tokens.sh
@@ -28,6 +28,9 @@ if [ ! -d "$TESTDIR/db" ]; then
 directories.tokendir = $TESTDIR/db
 objectstore.backend = file
 log.level = DEBUG
+# # The hashed ECDSA mechanisms wrongly do not support multi-part operations
+# https://github.com/softhsm/SoftHSMv2/issues/842
+slots.mechanisms = -CKM_ECDSA_SHA1,CKM_ECDSA_SHA224,CKM_ECDSA_SHA256,CKM_ECDSA_SHA384,CKM_ECDSA_SHA512
 EOF
 
     cat "$TESTDIR/softhsm.conf"
-- 
GitLab


